### PR TITLE
983 improve mobile support

### LINKED ISSUE #983: Improve mobile support
Test OPUS on a variety of phones and tablets and make it work as well as possible on all of them.

- Add message to "browser size" dialog suggesting the user rotates their phone.
- Fix problems with scrolling on search page left and right sides.
- Make work on smaller screens if at all possible.

### LINKED ISSUE #1005: Change font of "OR"
The "+OR" is in a serif font. Change it to sans serif to agree with the rest of OPUS.

### PR DESCRIPTION
**Pull request by juzen2003**
- Fixes #983 
- Fixes #1005 
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - Database used: opus3_test_200211
  - All Django tests pass: Y
- Were any JavaScript or CSS files modified? Y
  - JSHINT run on all affected files: Y 
  - Tested on Chrome, Firefox, Safari, and iOS: Y
    (Remember to test all browser sizes)
  - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
General: 
- Change supported browser size to 550px width & 270px height
- When browser width is less than 580px OR height is shorter than 400px:
    - Shrink the modal header 
    - Hide the footer
    - Set body font-size to be 75%
    - Set reset buttons font-size to be 90% (in search tab)
    - Shrink the height of main nav
- When browser width is less than 580px AND height is shorter than 400px:
    - Set the table header font size to 75% to have a better table display

Search:
- Recalculate mainNavHeight in searchBarContainerHeight & searchWidgetHeightChanged to have the correct height of search menu sidebar and widgets area.
- When browser width is less than 580px OR height is shorter than 630px:
    - Hide the feedback button
- When browser width is less than 580px OR height is shorter than 400px:
    - Reduce the padding around the whole search tab
    - Expand the width of widgets area

Browse:
- When browser height is shorter than 400px:
    - Hide download csv button in select metadata menu
- Update the logic in calculateGalleryHeight, take overlapped nav into consideration. This will remove the little gap between the end of gallery contents and the footer.

Cart:
- When browser height is shorter than 400px:
    - Hide product type select/deselect buttons in download panel
    - Hide "Download Options" in download panel
- When browser width is less than 580px OR height is shorter than 630px:
    - Hide download zip buttons in download panel
- When browser width is less than 580px OR height is shorter than 400px:
    - Reduce the top & bottom spaces around "Download Options" 
    - Reduce the space around data in download panel card body
    - Reduce the space around "Download Data" in download panel 

Detail:
- When browser height is less than 400px:
    - Keep the detail image stay at the right of the page
- When browser width is less than 768px AND height is shorter than 400px:
    - Make observation detail area shrinkable
    - Make observation detail area take up 70% of width and image take up 30%
    - Reduce the spaces around detail tab

Metadata modal for each observation:
- Make sure image always stays at the left in metadata modal
- Make sure metadata modal is not covered by main nav 
- Make sure hamburger menu is not covered by the modal

Known problems:
Mobile touch over-scrolling:
- In Pixel with Chrome (Firefox doesn't have this issue), in browse tab, there will be a huge gap at the end of gallery contents and footer.
- In iOS or Android devices with smaller screen (like iphone 8 or iphone se), in browse/cart tab, the 2nd navbar & gallery area can be over-scrolled. (The issue is not happened on iphone pro max).

### CODE REVIEW

**Comment by rfrenchseti** [comment]
- The screen is short enough that people are unlikely to want to download CSV files, how about removing "Download CSV (all results)" from the nav bar and moving the Observation # slider up in its place so there are only two rows of things under the main nav bar?
- The Select Metadata dialog is always a constant amount shorter than the browser height, even if the browser is really short. That's wasted space. How about allowing the Select Metadata dialog to be taller in that case?
- The padding around the buttons at the bottom of Select Metadata is really large on a small screen.
When the screen is short the padding at the top of the confirm modals goes away, but the padding below the title text doesn't, so it isn't vertically centered anymore.
- The Download Data slide-out actually works with much smaller browsers now, but the option to pull it out isn't available below a certain size.
- I think the Select all product types etc buttons should go away a little earlier when shrinking the browser because right now the scrollable at the bottom gets very short before the font changes and the buttons go away.
- The hamburger menu for thumbnails can go off the top or bottom of the screen. I think the font needs to be smaller / padding smaller when the browser is short.
- On the detail page, if you're looking at a VIMS/UVIS/CIRS observation, the text under the thumbnail that gives the link to the preview help changes font size around 370px high while the other text on the screen doesn't.
- The main menu is a little too tall for the screen when the browser is short. Maybe the font size or padding needs to be reduced.

**Comment by dstopp** [comment]
-the main nav and the padding around selected tab can be reduced 
-the height of the headers on the search widgets can be reduced a lot - they look oddly huge w/the changes
-the left and right padding on the header for select metadata is a little too small
-the padding between the header and the body of the select metadata modal is a little too big
-the buttons on the select metadata modal seem oddly large with respect to the rest of the page
-on the detail page, when the page is reduced to the absolute min, i think there is still a lot of extra space above "Observation Detail" title that can be reclaimed
-the size of the font awesome icons on the slideshow can be reduced, they look kind of odd. 
-the font for the pop up menu doesn't seem to be small, like the other fonts.  Do we want to remove the download CSV items from that menu?  
-should the slider number/box thing, font and padding, be reduced maybe 10-20%?  it still seems big, but OTOH i know we have to make it 'touchable' on a phone/small screen.

**Comment by juzen2003** [comment]
Updates:
- The 2nd nav in browse tab:
    - When screen width is less than 580 AND height is less than 400px:
        - Remove "Download CSV (all results)" from the nav bar and move the Observation # slider up
- Modal:
    - When screen width is less than 580 OR height is less than 400px:
        - Override .modal-header padding to top/bottom 0.25rem & left/right to 0.5rem 
        - Override .modal-body padding to top/bottom 0.25rem & left/right to 0.5rem 
        - Override .modal-footer padding to top/bottom 0.25rem & left/right to 0.5rem 
        - Reduce the modal buttons size
        - Keep the spaces for detail metadata modal in browse & cart tab to have the pretty display
        - Reduce the size of font awesome icon in slide show
- Select metadata menu:
    - When screen height is less than 400px: 
        - Increase the height of the menu by removing the 55px gap between the modal & bottom edge of the screen
- Cart download panel:
    - When screen width is less than 1100px OR height is less than 460px:
        - Cart download options pane will be hidden, and "Download Data" button will show up
    - When screen height is less than 400px:    
        - Move "Download Data" to the same as cart actions buttons
        - Remove the left pointing arrow
    - When screen width is less than 580px AND height is less than 400px:
        - Reduce spaces to make sure cart actions buttons are wrapped into the next line when screen is narrow & short
    - When screen height is less than 500px:
        - Hide select/deselect product types button
- Thumbnail hamburger menu:
    - When screen height is less than 400px:    
        - Reduce the spaces of each menu item in thumbnail hamburger
- Help menu: 
    - When screen height is less than 400px:    
        - Reduce the spaces of each menu item
        - Reduce the font size to 90%
- Detail tab:
    - Make sure the text below the image are the same size as ones in Observation detail
        - When screen height is less than 400px:
            - Reduce the spaces around "Observation Detail"
            - Reduce the spaces around "OPUS ID"
- Main nav:
    - When screen width is less than 580px OR height is less than 400px:
        - Reduce the padding around selected tab
        - Reduce the hamburger icon
- Widget:
    - When screen width is less than 580px OR height is less than 400px:
        - Reduce the padding around widget header
        - Reduce the spaces between each widget
- Slider:        
    - When screen width is less than 580 OR height is less than 400px:
        - Reduce the slider font size to 90%

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
Search Tab:

- As you narrow the browser, the Reset buttons don't stay centered under the horizontal line above them and there are a couple of strange twitches in horizontal position which seem like they could be smoothed out.
- The same twitch at browser width 700 and 705 cause word wrapping under Surface Geometry Constraints to flip back and forth. It seems like there should only be one break point instead of two.
- When the font changes size at 580, the widget titles don't change.
- Do you think it would be worth changing the font size somewhere larger, like 700, instead of 580? Below around 700 the search tab gets awfully crunched.
- This may be hard to do, but as the width drops below 970, the min/max fields go to two lines and the input fields fill the complete width of the widget. But the +OR or trashcan wrap to the next line. It seems like the input fields could be kept narrowed and the +OR/trashcan kept next to the second line longer. Something similar happens for string input fields. The qtype and +OR wrap long before it seems like they need to. The input field could be made narrower first.

Cart Tab:
- What is the reason for removing the product pane when the height is < 460?

Help Menu:
- We should remove "Contact Us" from the Help menu at the same time we remove the Feedback tab due to size.

Detail Tab:
- When the font size shrinks, I don't think the various category headers change size.

- How hard would it be to shrink the thumbnail size whenever we shrink the font size?

**Comment by dstopp** [comment]
-In the table view, the image produced by hover over the table image is too large when the screen size is reduced.
-On the modal, when screen width is just a bit greater than 580 and height is more than 400px, the icons are still too large and they overlap the metadata.

**Comment by rfrenchseti** [comment]
I'm assuming by "modal" Debby is referring to the slideshow view.

**Comment by juzen2003** [comment]
- General:
    - Change the breakpoint of shrinking font-size from 580 to 700px
    - When screen width is less than 700px OR height is less than 630px:
        - Hide "Contact Us" in help panel
- Search tab:
    - In browser resize event, call the non-debounced version of o_search.searchHeightChanged() to prevent the horizontal position twitch caused by the delayed calculation of search sidebar & widget height. 
    - When screen width is less than 700px OR height is less than 400px
        - When font shrinks, remove the top margin to make reset buttons horizontally centered
        - Reduce widget title font-size to 75%
- Cart tab:
    - Update the logic of calculating product type table height in adjustProductInfoHeight 
    - Keep the download data left pane when screen is short. Only hide it when screen width is less than 1100px
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the size of "Metadata CSV" button
        - Reduce the space around "Download Data" in download pane
- Detail tab:
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the font-size of category headers in observation detail
- Mini thumbnail in table view:
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the size of zoom-in mini-thumbnail on hover
- Select metadata menu:
    - When screen width is less than 700px:
        - Position "Download CSV (all results)" button right above the modal footer when screen is narrow
- Metadata modal:
    - Add some gap between font awesome icons and metadata info area
- Gallery thumbnail:
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the thumbnail container from 100x100 to 90x90 (10% off) when font size is changed
- STRING & RANGE input:
    - Set the width of input.STRING & input.RANGE to make sure it can be shrunk more when screen is shrinking (before reaching to the breakpoint at 940px)

- NOTE:
    - RE: twitch at browser width 700 and 705 cause word wrapping under Surface Geometry Constraints to flip back and forth. It seems like there should only be one break point instead of two.
        - The twitch of height change is caused by the nature of text wrapping, we didn't have any breakpoints around 700 & 705px width. For text wrapping, we can only set from one of these: break by word, break by letter, or no break at all. Break by word is the default, and the twitch here is caused by the default nature. I don't think we can prevent this because we need the text wrap by word.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- The Feedback tab / contact us disappears at 700px width. It doesn't need to go away until 560px which is what I thought the previous behavior was.
- In the cart download pane, the "download archive" buttons go away at height 630px, and the "download CSV' button is still here. I don't see any reason for the archive/URL buttons to go away. They also go away at height 700px. Frankly I don't see any reason the buttons ever need to go away.
- In the slideshow, I think there's too much vertical space between the bottom icons and the bottom of the metadata container.
- At width 1100px, there's a weird change in the width of the string and range input fields. I don't think the change needs to happen there because there is still plenty of whitespace to the right of the fields. The browser could be narrower before shrinking the inputs.
- I don't think we were talking about the same "twitch" around 700. Whatever it was, it's gone now though.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- On the iPhone, in table view, click the "add all to cart" icon. Most of the time the modal verification dialog will show up at a weird z-layer behind the table header. Sometimes it won't show up at all.
- On the iPhone, in table view, click the hamburger then "start range". The green square will appear but won't be animated. It seems to work fine on the gallery view, or if you start the range on the gallery and then switch to the table. Sounds like it might be a browser bug.

**Comment by juzen2003** [comment]
- Feedback & contact us:
    - When screen width is less than 560px OR height is less than 630px:
        - Hide feedback tab & contact us option in help menu
- Data Archive & URL Archive buttons in download pane:
    - Both buttons will not go away now
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the spaces around both buttons
- Metadata modal:
    - Reduce the gap between the bottom icons and the container of metadata info
- Input.RANGE: 
    - Remove the width styling at 1100px breakpoint
- "Add All Results to Cart" table header icon:
    - Set title attributes to the button tag that wraps the cart icon (i tag). This will make sure default tooltip is properly displayed on PC, and mobile device touch start will trigger the button click event handler instead of the icon touchstart event handler
- RE: green square animation won't run when clicking thumbnail at table view. This happened in Safari on both iphone & mac. If we manually update or remove & add the -webkit-animation: hvr-ripple-in 1s infinite; styling for .hvr-ripple-in::before in js console, the animation will run again. However, when we modify it using js (the styling value is successfully modified in js console), there will be still no animation. Here is what I did in css & js:
css:
```
.hvr-ripple-in::before {
    -webkit-animation: var(--webkit-animation, hvr-ripple-in 1s infinite);;
}
```
js: (at the end of highlightStartOfRange, update the --webkit-animation)
```
$(".hvr-ripple-in").css("--webkit-animation", "hvr-ripple-in 1.1s infinite");
$(".hvr-ripple-in").css("--webkit-animation", "hvr-ripple-in 1s infinite");
```

File: opus/application/static_media/css/opus.css [hunk 1]
```diff
@@ -259,7 +260,8 @@ Not sure if this is the desired behavior
 
 /* As long as q-type shrink to the 2nd line, we set text input width
 to 100%, this will make the shrinking look smoother */
-@media (max-width: 970px) {
+@media (max-width: 940px) {
+/* @media (max-width: 970px) { */
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Remove comment

File: opus/application/static_media/css/opus.css [hunk 2]
```diff
@@ -1373,7 +1373,7 @@ to stay in one line */
 
     /* Move the last item in flexbox to the 2nd line */
     #browse .browse-nav-container .navbar-nav {
-        flex-wrap: wrap;
+        /* flex-wrap: wrap; */
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Remove comment

File: opus/application/static_media/css/opus.css [hunk 3]
```diff
@@ -1577,6 +1562,7 @@ to stay in one line */
 
 /* Remove the nav-link for opening up download data panel when screen
 is wide enough */
+/* @media (min-width: 1101px) and (min-height: 461px) { */
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Remove comment

File: opus/application/static_media/css/opus.css [hunk 4]
```diff
@@ -2764,3 +2757,366 @@ when wrapping into the 2nd line */
     padding-left: 0.2rem;
     padding-right: 0.2rem;
 }
+
+/* The followings are for mobile support (width: 550px and height: 270px) */
+@media (max-width: 700px), (max-height: 400px){
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Add space before {

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- When the browser is short enough that the footer goes away, and then you start add/remove range, the text bubble that pops up in the lower left is now too tall. I think it's assuming that the footer is still there.
- In table view, when you start add/remove range and the green square is pulsing, if you scroll the table up that particular line entry will actually move over the table headers. It has the wrong Z value somehow. This happens on all browsers.
- I still don't see the square pulsing in the table view on the iPhone. I also don't see the changes you made in the actual code. Did you forget a push?

**Comment by juzen2003** [comment]
Updates:
- Range select info box at lower left corner:
    - When screen width is less than 700px OR height is less than 400px:
        - When footer is gone, make sure the range select info box has the proper height
- hvr-ripple-in:
    - Set z-index to auto to avoid overlapping with table header

Known issue:
On Safari, the select range green pulse animation for pseudo element (hvr-ripple-in::before) in browse table view will not be triggered. This only happen on browse table view on Safari. The manually workaround to restart the animation is to unselect & select the following stylings for hvr-ripple-in::before in Safari dev tool:
  - -webkit-animation: hvr-ripple-in 1s infinite;
  - animation: hvr-ripple-in 1s infinite;

However remove & add the stylings by codes will not restart the animation. So we don't have a fix for this one.

**Comment by rfrenchseti** [comment]
Actually the green-box-not-moving problem also seems to happen with Chrome on iOS.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
There is still one really fatal problem with small browsers. It now doesn't count the number of thumbnails properly. If you set your browser to the size of an iPhone 7 (which has 7 thumbnails across), the slider will only increment by 6. This makes a mess of everything else.

**Comment by juzen2003** [comment]
- Update to change o_browse.imageSize from 100 to 90 when browser width is less than 700px or height is less than 400px. The values 100 & 90 need to match the min-width & min-height values for .op-thumbnail-container in opus.css.

**Review by dstopp — APPROVED** [review]
looks good; note that merge of 362 will likely have a lot of conflicts..

**Review by rfrenchseti — APPROVED** [review]

**Pull request merged**